os/arch/arm/src/common/up_allocateheap.c

```diff
@@ -129,14 +129,16 @@ void up_allocate_kheap(FAR void **heap_start, size_t *heap_size)
 	/* Return the kernel heap settings (i.e., the part of the heap region
 	 * that was not dedicated to the user heap).
 	 */
-
-	*heap_start = (FAR void *)(g_idle_topstack & ~(0x7));
-
-	/* There may be a special scenario where we might configure a different region
-	 * for heap. In such case, if end of bss falls outside of the region address range,
-	 * then we use the whole region for heap.
-	 */
-	if (*heap_start < (void *)KREGION_START || *heap_start > (void *)KREGION_END) {
+	void *stack_end = (void *)(g_idle_topstack & ~(0x7));
+
+	if ((void *)&_sdata >= (void *)KREGION_START && (void *)&_sdata <= (void *)KREGION_END && stack_end >= (void *)KREGION_END) {
```

Comment:
```c
if ((void *)KREGION_START < stack_end && (void *)KREGION_END > (void *)&_stext) {
```
I am sorry but I don't know why you check it with _sdata. 
Usually _stext is the start of flash layout and stack_end is the last so that I think above covers all of cases.

Reply:
As you mentioned, stext is for flash region. We need to check for ram region. So, I am checking with sdata. 
I am checking if data section starts within current ram region and also the stack ends within the current ram region. If this condition is not satisfied, then it is an error.

Reply:
Oh ok. You're right. We should check it with sdata, not stext.
Then
```
(void *)&_sdata <= (void *)KREGION_END && stack_end >= (void *)KREGION_END
```
Isn't this enough?
```
(void *)&_sdata >= (void *)KREGION_START
```
Is this necessary?
If we add this conditional, we can't check the case entire kregion is used (overlapped).
Layout like below:
`_sdata - KREGION_START - KREGION_END - stack_end`

Reply:
Yes. You are right. I modified it.